packages/flutter_tools/lib/src/artifacts.dart

```diff
@@ -162,6 +164,8 @@ String? _artifactToFileName(Artifact artifact, [ TargetPlatform? platform, Build
   switch (artifact) {
     case Artifact.genSnapshot:
       return 'gen_snapshot';
+    case Artifact.libFlutterEngine:
+      return 'libflutter_engine.so';
```

Comment:
Is this file part of the officially pre-built Flutter SDK, or only generated via a custom device engine build?

For example, on x64 linux I have `libflutter_linux_gtk.so`.

Reply:
I hope I am grasping your question fully. 

**To give some background first:**

If build the flutter engine locally with `./flutter/tools/gn --target-os linux --linux-cpu x64` then the `out/linux_debug_x64` folder has `libflutter_engine.so` and `libflutter_linux_gtk.so` , I am not 100% sure what the difference between the two is but I do know that having`libflutter_engine.so` available is very convenient when working with a custom device. 

I strongly suspect that `libflutter_engine.so` is "always" generated when building the engine. 

My python is not great but from what I read [here ](https://flutter.googlesource.com/recipes/+/refs/heads/master/recipes/engine.py) in the function starting on ln754 `libflutter_engine.so` is plainly not copied to the hosted artifacts folder.

This pull request was written with locally built engines in mind. My reasoning is that working with a custom device would imply that you are building your own engine. Since if it was hosted by infra it would not be a custom device. 

**Answer:**

As far as my knowledge goes `libflutter_linux_gtk.so` would be downloaded from infra and is not part of the SDK. 

Assuming that officially pre-built flutter SDK is referring to engine build artifacts hosted infra, my answer is : No , it not part of the officially pre-built Flutter SDK , and is generated via a linux engine build.

Reply:
This should be whatever the output of the `flutter_engine_library` target is: https://github.com/flutter/engine/blob/main/shell/platform/embedder/BUILD.gn#L291.

That will be toolchain dependent. On Linux it'll be `libflutter_engine.so`.

Reply:
Thanks !

Reply:
I am a bit beat on how to solve this problem in a clean fashion, solutions I can think of :

1. Have a "hardcoded" lookup of some sort that returns different names based on the value of `platform (ln 160).

2. Have `engineLibraryFileName` as part of [custom_devices.json](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/static/custom-devices.schema.json) and use that to define the artifact file name.

3. A combination of the above options where we use 2 and have a "hardcoded" fallback or two.

Reply:
```suggestion
      r      switch (platform) {
        case TargetPlatform.windows_x64:
        case TargetPlatform.windows_uwp_x64:
          return 'flutter_engine.dll';
        case TargetPlatform.android:
        case TargetPlatform.android_arm:
        case TargetPlatform.android_arm64:
        case TargetPlatform.android_x64:
        case TargetPlatform.android_x86:
          throw StateError('Artifact $artifact not available for platform $platform.');
        case TargetPlatform.darwin:
          return 'libflutter_engine.dylib';
        case TargetPlatform.fuchsia_arm64:
        case TargetPlatform.fuchsia_x64:
        case TargetPlatform.ios:
          throw StateError('Artifact $artifact not available for platform $platform.');
        case TargetPlatform.linux_arm64:
        case TargetPlatform.linux_x64:
          return 'libflutter_engine.so';
        case TargetPlatform.tester:
        case TargetPlatform.web_javascript:
        case null:
          throw StateError('Artifact $artifact not available for platform $platform.');
      }
```